staging/src/k8s.io/endpointslice/reconciler_test.go

```diff
@@ -1972,6 +1973,197 @@ func TestReconcileTopology(t *testing.T) {
 	}
 }
 
+// Test reconciliation behaviour for trafficDistribution field.
+func TestReconcile_TrafficDistribution(t *testing.T) {
+	// Setup the following topology for the test.
+	//
+	// 	- node-0 IN zone-a CONTAINS {pod-0}
+	// 	- node-1 IN zone-b CONTAINS {pod-1, pod-2, pod-3}
+	// 	- node-2 IN zone-c CONTAINS {pod-4, pod-5}
+	ns := "ns1"
+	svc, _ := newServiceAndEndpointMeta("foo", ns)
+	nodes := []*corev1.Node{}
+	pods := []*corev1.Pod{}
+	for i := 0; i < 3; i++ {
+		nodes = append(nodes, &corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("node-%v", i),
+				Labels: map[string]string{
+					corev1.LabelTopologyZone: fmt.Sprintf("zone-%c", 'a'+i),
+				},
+			},
+			Status: corev1.NodeStatus{
+				Conditions: []corev1.NodeCondition{{
+					Type:   corev1.NodeReady,
+					Status: corev1.ConditionTrue,
+				}},
+				Allocatable: corev1.ResourceList{"cpu": resource.MustParse("100m")},
+			},
+		})
+	}
+	for i := 0; i < 6; i++ {
+		pods = append(pods, newPod(i, ns, true, 1, false))
+	}
+	pods[0].Spec.NodeName = nodes[0].Name
+	pods[1].Spec.NodeName = nodes[1].Name
+	pods[2].Spec.NodeName = nodes[1].Name
+	pods[3].Spec.NodeName = nodes[1].Name
+	pods[4].Spec.NodeName = nodes[2].Name
+	pods[5].Spec.NodeName = nodes[2].Name
+
+	// Define test cases.
+
+	testCases := []struct {
+		name string
+		desc string
+
+		trafficDistributionFeatureGateEnabled bool
+		trafficDistribution                   string
+		topologyAnnotation                    string
+
+		// Defines how many hints belong to a particular zone.
+		wantHintsDistributionByZone map[string]int
+		wantMetrics                 expectedMetrics
+	}{
+		{
+			name:                                  "trafficDistribution=PreferClose, topologyAnnotation=Disabled",
+			desc:                                  "When trafficDistribution is enabled and topologyAnnotation is disabled, hints should be distributed as per the trafficDistribution field",
+			trafficDistributionFeatureGateEnabled: true,
+			trafficDistribution:                   corev1.ServiceTrafficDistributionPreferClose,
+			topologyAnnotation:                    "Disabled",
+			wantHintsDistributionByZone: map[string]int{
+				"zone-a": 1, // {pod-0}
+				"zone-b": 3, // {pod-1, pod-2, pod-3}
+				"zone-c": 2, // {pod-4, pod-5}
+			},
```

Comment:
Why not explicitly check that the expected pods have the expected zones? The current tests could let a bug through where the correct number of pods were allocated, just the wrong ones, and thus you'd still end up with a lot of cross-zone traffic.